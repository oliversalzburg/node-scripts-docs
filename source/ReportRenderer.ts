import { FragmentStore } from "./FragmentStore.js";
import { ScriptStore } from "./ScriptStore.js";
import { ValidationReport } from "./Validator.js";

export const render = (report: ValidationReport, skipPending = true) => {
  const output = new Array<string>();
  if (0 < report.pendingDocumentation.size && !skipPending) {
    output.push(" --- Pending documentation --- ");
    for (const scriptMeta of report.pendingDocumentation) {
      output.push(
        `  ${ScriptStore.makeScriptLocator(scriptMeta.projectName, scriptMeta.scriptName)}`,
      );
    }
    output.push("");
  }

  if (0 < report.missingFragments.size) {
    output.push(" --- Missing fragments (will be generated) --- ");
    for (const scriptMeta of report.missingFragments) {
      output.push(
        `  ${FragmentStore.scriptToFragmentFilename(scriptMeta.scriptName)} (${
          scriptMeta.scriptName
        })`,
      );
    }
    output.push("");
  }

  if (0 < report.corruptedMetadataRecords.size) {
    output.push(" --- Outdated metadata (will be updated) --- ");
    for (const scriptMeta of report.corruptedMetadataRecords) {
      output.push(
        `  ${ScriptStore.makeScriptLocator(scriptMeta.projectName, scriptMeta.scriptName)}`,
      );
    }
    output.push("");
  }

  if (0 < report.obsoleteFragments.size) {
    output.push(" --- Obsolete fragments (delete manually) --- ");
    for (const fragment of report.obsoleteFragments) {
      output.push(`  ${fragment.filename}`);
    }
    output.push("");
  }

  if (0 < report.changedFragments.size) {
    output.push(" --- Detected changes --- ");
    for (const scriptMeta of report.changedFragments) {
      output.push(
        `  ${FragmentStore.scriptToFragmentFilename(scriptMeta.scriptName)} (${
          scriptMeta.scriptName
        })`,
      );
    }
    output.push("");
  }

  if (0 < report.newScripts.size) {
    output.push(" --- New scripts --- ");
    for (const scriptMeta of report.newScripts) {
      output.push(
        `  ${ScriptStore.makeScriptLocator(scriptMeta.projectName, scriptMeta.scriptName)}`,
      );
    }
    output.push("");
  }

  return output.join("\n");
};
