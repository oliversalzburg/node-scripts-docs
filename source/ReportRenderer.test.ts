import { render } from "./ReportRenderer.js";

it("renders empty report as expected", () => {
  const report = {
    changedFragments: new Set([]),
    corruptedMetadataRecords: new Set([]),
    missingFragments: new Set([]),
    newScripts: new Set([]),
    obsoleteFragments: new Set([]),
    pendingDocumentation: new Set([]),
    unchangedFragments: new Set([]),
  };

  expect(render(report, false)).toMatchSnapshot();
  expect(render(report)).toMatchSnapshot();
});

it("renders report as expected", () => {
  const report = {
    changedFragments: new Set([
      {
        descriptiption: "something useful",
        isGlobal: false,
        isRootManifest: false,
        manifestPath: "test/fixtures/default/package.json",
        projectName: "default",
        scriptCode: "echo nothing",
        scriptName: "changed",
      },
    ]),
    corruptedMetadataRecords: new Set([
      {
        descriptiption: "something useful",
        isGlobal: false,
        isRootManifest: false,
        manifestPath: "test/fixtures/default/package.json",
        projectName: "default",
        scriptCode: "echo nothing",
        scriptName: "corrupted",
      },
    ]),
    missingFragments: new Set([
      {
        descriptiption: "something useful",
        isGlobal: false,
        isRootManifest: false,
        manifestPath: "test/fixtures/default/package.json",
        projectName: "default",
        scriptCode: "echo nothing",
        scriptName: "missing",
      },
    ]),
    newScripts: new Set([
      {
        descriptiption: "something useful",
        isGlobal: false,
        isRootManifest: false,
        manifestPath: "test/fixtures/default/package.json",
        projectName: "default",
        scriptCode: "echo nothing",
        scriptName: "new",
      },
    ]),
    obsoleteFragments: new Set([
      {
        descriptionMarkdown: "something useful",
        filename: ".obsolete.md",
      },
    ]),
    pendingDocumentation: new Set([
      {
        descriptiption: "something useful",
        isGlobal: false,
        isRootManifest: false,
        manifestPath: "test/fixtures/default/package.json",
        projectName: "default",
        scriptCode: "echo nothing",
        scriptName: "pending",
      },
    ]),
    unchangedFragments: new Set([
      {
        descriptiption: "something useful",
        isGlobal: false,
        isRootManifest: false,
        manifestPath: "test/fixtures/default/package.json",
        projectName: "default",
        scriptCode: "echo nothing",
        scriptName: "unchanged",
      },
    ]),
  };

  expect(render(report, false)).toMatchSnapshot();
  expect(render(report)).toMatchSnapshot();
});
