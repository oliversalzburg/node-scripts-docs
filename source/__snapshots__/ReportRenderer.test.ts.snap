// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`renders empty report as expected 1`] = `""`;

exports[`renders empty report as expected 2`] = `""`;

exports[`renders report as expected 1`] = `
" --- Pending documentation --- 
  pending→default

 --- Missing fragments (will be generated) --- 
  .missing.md (missing)

 --- Outdated metadata (will be updated) --- 
  corrupted→default

 --- Obsolete fragments (delete manually) --- 
  .obsolete.md

 --- Detected changes --- 
  .changed.md (changed)

 --- New scripts --- 
  new→default
"
`;

exports[`renders report as expected 2`] = `
" --- Missing fragments (will be generated) --- 
  .missing.md (missing)

 --- Outdated metadata (will be updated) --- 
  corrupted→default

 --- Obsolete fragments (delete manually) --- 
  .obsolete.md

 --- Detected changes --- 
  .changed.md (changed)

 --- New scripts --- 
  new→default
"
`;
